### ISSUE TITLE
User Model

### ISSUE THREAD

**Issue by WorldWideWebb** [wells/l4-ldap-ntlm#6]
Any plans on baking in (or an example, if you can already) some clean functionality for using a User model (to store additional user attributes in the database)? Works great otherwise!

**Comment by wells** [comment]
Yes. I have plans on a user model to help with tokens. Plus a db to handle named route based RW permissions for LDAP groups and users.

**Comment by WorldWideWebb** [comment]
Sweet. Thanks.

**Comment by nesl247** [comment]
Any update to this? Really need this functionality.

**Comment by WorldWideWebb** [comment]
I made a hack to do this. I made a blank "user" object as a config property. So I can access user attributes anywhere in the app by:

sails.config.user.anypropertynamehere

I add any attributes I want to store for my users normally (as part of the user model). Once the user logs in (using passport to auth against active directory), I update the database (just like I normally would any other model) with the properties that I want to store locally (for example, I grab their first and last name and email address), then grab that user from the database (again, like you normally would query a model), and save that user to the config user object I created. In the logout action, I also delete that config user object, because passport (I'm guessing it's passport that's doing this) doesn't seem to fully clear out the session when you use the logout method.

Like I said, kind of hacky, but it seems to work well. Let me know if you want code samples.

**Comment by nesl247** [comment]
Before I do that, I'll probably just fork this and correct it so that it stops returning an instance of `GenericUser`. The https://github.com/ccovey/ldap-auth library does it this way (merges LDAP and the user model). The problem with that library is that it doesn't work with 4.2. It's a lose, lose situation all around.

**Comment by wells** [comment]
I will implement what I've been talking about, but it has been sidelined until I'm caught up to speed with all of my other projects.

**Comment by nesl247** [comment]
Looking forward to your implementation. I'm not sure if what you are planning is the same as what is in ccovey\ldap-auth. All his model integration does is tie into a model when using `Auth::user()`. Unfortunately getting an instance of the model itself for any other user does not include the LDAP integration.

If possible, please include support for making the model itself mixed with LDAP, and not just when accessing the `Auth::user()` instance.

**Comment by wells** [comment]
@WorldWideWebb @nesl247 So far, my User model will have username, email, domain, perhaps an hr_id, and remember_token.

Also @nesl247, do you want the LDAP fields to be loaded into the User model or accessed through a function, such as `$user->ldap()->attribute;` I could even have a subset loaded into the User model and have the remaining attributes be accessible via a function call.

**Comment by nesl247** [comment]
@wells The way I would imagine it is to always retrieve all fields, unless a `fields` array is specified. Store everything inside of a `$ldapAttributes` property. Override the Eloquent magic methods so that it checks for both `$attributes` and `$ldapAttributes`. This way everything is accessed just like it always has been.

The thing that I really would like though is I want this to be for all instances of `User`, not just for the logged in user. Therefore we would need a new class to extend, instead of Eloquent. This class would obviously extend Eloquent itself.

The reason for this is that way we don't have to maintain a bunch of these fields inside of the database, just to be able to access the information for users outside of the logged in one.

**Comment by rkgarcia** [comment]
With a little modification we can add things like permissions
`
class User extends GenericUser {

	public function __construct( $items )
	{
		parent::__construct( $items );
		$this->permissions = []; /* permissions query here bassed upon group */
	}
}
`
